### PR TITLE
Zipformer for TedLium

### PR DESCRIPTION
**Pull request by desh2608**
This is a TedLium recipe for the new zipformer model. I am still tuning the training recipe so this is WIP.

### CODE REVIEW

**Comment by desh2608** [comment]
The first model trained well (https://tensorboard.dev/experiment/KKmyNRRjRgqL7J28lNq7EA/#scalars) but the WER is worse than other tedlium3 models:

|                      | dev   | test  |
|----------------------|-------|-------|
| greedy_search        | 10.58 | 11.76 |
| modified_beam_search | 13.47 | 15.02 |

In particular, I found that the insertion error rate was much higher than the other tedlium3 models. I think this may be because of the issue fixed in [this PR](https://github.com/k2-fsa/icefall/pull/942). I have made this change now and training another model.

**Comment by desh2608** [comment]
> The first model trained well (https://tensorboard.dev/experiment/KKmyNRRjRgqL7J28lNq7EA/#scalars) but the WER is worse than other tedlium3 models:
> 
> dev	test
> greedy_search	10.58	11.76
> modified_beam_search	13.47	15.02
> In particular, I found that the insertion error rate was much higher than the other tedlium3 models. I think this may be because of the issue fixed in [this PR](https://github.com/k2-fsa/icefall/pull/942). I have made this change now and training another model.

It seems that even after making this change, the resulting model has a similar WER. As a comparison, the conformer-CTC model gets ~6-7% WER. Any suggestions what may be happening?

**Comment by marcoyang1998** [comment]
Based on the WER table, it seems that using `modified_beam_search` produces higher WER than `greedy_search`. This is a bit strange.

**Comment by videodanchik** [comment]
Hey @desh2608, If the most insertions are `<unk>` keep in mind that there are a lot of `<unk>` in the train data, which most probably refers to punctuation. So my advice to either remove these `<unk>` from train or explicitly remove it from ASR decoding output.

**Comment by desh2608** [comment]
There are a lot of errors in the beginning of the utterance. For example:

```
AlGore_2009-1:	ref=['last', 'year', 'i', 'showed', 'these', 'two', 'slides', 'so', 'that', 'demonstrate', 'that', 'the', 'arctic', 'ice', 'cap', 'which', 'for', 'most', 'of', 'the', 'last', 'three', 'million', 'years', 'has', 'been', 'the', 'size', 'of', 'the', 'lower', 'forty', 'eight', 'states', 'has', 'shrunk', 'by', 'forty', 'percent']
AlGore_2009-1:	hyp=['⁇', 'unk', '⁇', 'these', 'two', 'slides', 'that', 'demonstrate', 'that', 'the', 'arctic', 'ice', 'cap', 'which', 'for', 'most', 'of', 'the', 'last', 'three', 'million', 'years', 'has', 'been', 'the', 'size', 'of', 'the', 'lower', 'forty', 'eight', 'states', 'has', 'shrunk', 'by', 'forty', 'percent']
```

**Comment by videodanchik** [comment]
@desh2608 Yes this is the common pattern `'⁇', 'unk', '⁇'` so you experiencing errors related to bad tokenization of these `<unk>`. So I recommend you to look into my recipe with a reworked conformer model, I did some analyses of these `<unk>` in the recipe https://github.com/k2-fsa/icefall/pull/696

**Comment by desh2608** [comment]
> @desh2608 Yes this is the common pattern `'⁇', 'unk', '⁇'` so you experiencing errors related to bad tokenization of these `<unk>`. So I recommend you to look into my recipe with a reworked conformer model, I did some analyses of these `<unk>` in the recipe #696

Thanks for the pointer. This discussion looks very useful!

**Comment by desh2608** [comment]
After fixing the <unk> decoding as per suggestion from @videodanchik, the resulting model is getting reasonable WERs (~6-7%). I am tuning it a bit more to improve the WER.

**Comment by csukuangfj** [comment]
> Here are the final results:
> 
> dev	test	comment
> greedy search	6.74	6.16	--epoch 50, --avg 22, --max-duration 500
> beam search (beam size 4)	6.56	5.95	--epoch 50, --avg 22, --max-duration 500
> modified beam search (beam size 4)	6.54	6.00	--epoch 50, --avg 22, --max-duration 500
> fast beam search (set as default)	6.91	6.28	--epoch 50, --avg 22, --max-duration 500
> I think this is ready to merge now.

Sorry for missing this comment. Will review it tomorrow.

**Comment by videodanchik** [comment]
@desh2608 @csukuangfj I also wonder if `--ref-duration` argument in `train.py` should be adjusted accordingly, when `--max-duration` is reduced to 500? How to choose this value correctly? Should it be ~80-90% of `--max-duration`?

**Review by yaozengwei — APPROVED** [review]

**Comment by yaozengwei** [comment]
> Here are the final results:
> 
> dev	test	comment
> greedy search	6.74	6.16	--epoch 50, --avg 22, --max-duration 500
> beam search (beam size 4)	6.56	5.95	--epoch 50, --avg 22, --max-duration 500
> modified beam search (beam size 4)	6.54	6.00	--epoch 50, --avg 22, --max-duration 500
> fast beam search (set as default)	6.91	6.28	--epoch 50, --avg 22, --max-duration 500
> I think this is ready to merge now.

It's a little strange that fast-beam-search gets worse result compared to greedy-search.

**Comment by csukuangfj** [comment]
> 
> 
> 
> 
> 
> 
> > Here are the final results:
> 
> > 
> 
> > dev	test	comment
> 
> > greedy search	6.74	6.16	--epoch 50, --avg 22, --max-duration 500
> 
> > beam search (beam size 4)	6.56	5.95	--epoch 50, --avg 22, --max-duration 500
> 
> > modified beam search (beam size 4)	6.54	6.00	--epoch 50, --avg 22, --max-duration 500
> 
> > fast beam search (set as default)	6.91	6.28	--epoch 50, --avg 22, --max-duration 500
> 
> > I think this is ready to merge now.
> 
> 
> 
> It's a little strange that fast-beam-search gets worse result compared to greedy-search. 

Perhaps Desh is using an old version of k2?

**Comment by desh2608** [comment]
```
Collecting environment information...

k2 version: 1.24.3
Build type: Debug
Git SHA1: 38211604d6a24b15f320578a1a38f6c12d7a711c
Git date: Mon Jun 12 10:59:44 2023
Cuda used to build k2: 11.7
cuDNN used to build k2:
Python version used to build k2: 3.8
OS used to build k2: CentOS Linux release 7.5.1804 (Core)
CMake version: 3.18.6
GCC version: 7.2.0
CMAKE_CUDA_FLAGS:  --compiler-options -rdynamic --compiler-options -lineinfo -Wno-deprecated-gpu-targets   -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_35,code=sm_35  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_50,code=sm_50  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_60,code=sm_60  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_61,code=sm_61  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_70,code=sm_70  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_75,code=sm_75  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_80,code=sm_80  -lineinfo --expt-extended-lambda -use_fast_math -Xptxas=-w  --expt-extended-lambda -gencode arch=compute_86,code=sm_86 -DONNX_NAMESPACE=onnx_c2 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_80,code=sm_80 -gencode arch=compute_86,code=sm_86 -gencode arch=compute_86,code=compute_86 -Xcudafe --diag_suppress=cc_clobber_ignored,--diag_suppress=integer_sign_change,--diag_suppress=useless_using_declaration,--diag_suppress=set_but_not_used,--diag_suppress=field_without_dll_interface,--diag_suppress=base_class_has_different_dll_interface,--diag_suppress=dll_interface_conflict_none_assumed,--diag_suppress=dll_interface_conflict_dllexport_assumed,--diag_suppress=implicit_return_from_non_void_function,--diag_suppress=unsigned_compare_with_zero,--diag_suppress=declared_but_not_referenced,--diag_suppress=bad_friend_decl --expt-relaxed-constexpr --expt-extended-lambda -D_GLIBCXX_USE_CXX11_ABI=0 --compiler-options -Wall  --compiler-options -Wno-strict-overflow  --compiler-options -Wno-unknown-pragmas
CMAKE_CXX_FLAGS:  -D_GLIBCXX_USE_CXX11_ABI=0 -Wno-unused-variable  -Wno-strict-overflow
PyTorch version used to build k2: 2.0.0+cu117
PyTorch is using Cuda: 11.7
NVTX enabled: True
With CUDA: True
Disable debug: False
Sync kernels : True
Disable checks: True
Max cpu memory allocate: 214748364800 bytes (or 200.0 GB)
k2 abort: False
__file__: /exp/draj/jsalt2023/k2/k2/python/k2/version/version.py
_k2.__file__: /exp/draj/jsalt2023/k2/build_debug/lib/_k2.cpython-38-x86_64-linux-gnu.so
```
I think it's pretty new?

I think the likely source of error is that `allow_partial` is set to False by default in the fast_beam_search method. I see I have higher deletion error than greedy search, which could explain this.

**Comment by desh2608** [comment]
Even after using `allow_partial=True`, I get the same WER. 

BTW, I added this option (set to `False` by default) in fast_beam_search related methods in beam_search.py.

**Comment by csukuangfj** [comment]
Thanks!

**Comment by desh2608** [comment]
BTW I found in my recent experiments that using `rnnt_type="modified"` provides significant improvements in WER for Tedlium (see RESULTS file). @videodanchik FYI

**Comment by pkufool** [comment]
> BTW I found in my recent experiments that using `rnnt_type="modified"` provides significant improvements in WER for Tedlium (see RESULTS file). @videodanchik FYI

What's your decoding method,  In my previous experiments,  `rnnt_type="modified"` gave very poor results when decoding with greedy search.

**Comment by desh2608** [comment]
@pkufool it was better than the regular topology on all decoding methods. See the tables below for comparison:

**rnnt_type="regular"**

|                                    |     dev    |    test    | comment                                  |
|------------------------------------|------------|------------|------------------------------------------|
|          greedy search             | 6.74       | 6.16       | --epoch 50, --avg 22, --max-duration 500 |
| modified beam search (beam size 4) | 6.54       | 6.00       | --epoch 50, --avg 22, --max-duration 500 |
| fast beam search (set as default)  | 6.91       | 6.28       | --epoch 50, --avg 22, --max-duration 500 |

**rnnt_type="modified"**

|                                    |     dev    |    test    | comment                                  |
|------------------------------------|------------|------------|------------------------------------------|
|          greedy search             | 6.32       | 5.83       | --epoch 50, --avg 22, --max-duration 500 |
| modified beam search (beam size 4) | 6.16       | 5.79       | --epoch 50, --avg 22, --max-duration 500 |
| fast beam search (set as default)  | 6.30     | 5.89       | --epoch 50, --avg 22, --max-duration 500 |

**Comment by danpovey** [comment]
@desh2608 can you see whether there a big differences in ins/del ratios, and whether there are particular types of words that are most affected.  Also try to compare your tensorflow plots for the 2 expts, so we can see if anything anomalous is happening.

BTW I think our modified RNN-T is equivalent to monotonic RNN-T: https://ieeexplore.ieee.org/document/9003822

**Comment by pkufool** [comment]
@desh2608  Ah I might have mixed up the results, I think you were training with `lm_scale` and decoding with `max-symbol-per-frame=1`. Here are the results from our paper. The model trained without `lm_scale` and decoded with `max-symbol-per-frame>1` get very poor results (a lost of insertions).
image

File: egs/tedlium3/ASR/RESULTS.md [hunk 1]
```diff
@@ -1,5 +1,133 @@
...
+  --max-duration 1000 \
+  --rnnt-type modified
+```
+
+|                                    |     dev    |    test    | comment                                  |
+|------------------------------------|------------|------------|------------------------------------------|
+|          greedy search             | 6.32       | 5.83       | --epoch 50, --avg 22, --max-duration 500 |
+|      beam search (beam size 4)     |        |        | --epoch 50, --avg 22, --max-duration 500 |
+| modified beam search (beam size 4) | 6.16       | 5.79       | --epoch 50, --avg 22, --max-duration 500 |
+| fast beam search (set as default)  | 6.30    ß   | 5.89       | --epoch 50, --avg 22, --max-duration 500 |
```
Comment (pkufool) [inline egs/tedlium3/ASR/RESULTS.md:105]: No results for `modified beam search` and an extra ` ß` in `fast beam search` row.

File: egs/tedlium3/ASR/RESULTS.md [hunk 2]
```diff
@@ -1,5 +1,133 @@
...
+```
+
+|                                    |     dev    |    test    | comment                                  |
+|------------------------------------|------------|------------|------------------------------------------|
+|          greedy search             | 6.58       | 6.20       | --epoch 50, --avg 22, --max-duration 500 |
+|      beam search (beam size 4)     | 6.34       | 5.92       | --epoch 50, --avg 22, --max-duration 500 |
+| modified beam search (beam size 4) | 6.38       | 5.84       | --epoch 50, --avg 22, --max-duration 500 |
+| fast beam search (set as default)  | 6.68       | 6.29       | --epoch 50, --avg 22, --max-duration 500 |
+
+A pre-trained model and decoding logs can be found at .
```
Comment (pkufool) [inline egs/tedlium3/ASR/RESULTS.md:129]: No download link here.

File: egs/tedlium3/ASR/RESULTS.md [hunk 3]
```diff
@@ -1,5 +1,133 @@
...
+```
+
+|                                    |     dev    |    test    | comment                                  |
+|------------------------------------|------------|------------|------------------------------------------|
+|          greedy search             | 6.32       | 5.83       | --epoch 50, --avg 22, --max-duration 500 |
+|      beam search (beam size 4)     |        |        | --epoch 50, --avg 22, --max-duration 500 |
+| modified beam search (beam size 4) | 6.16       | 5.79       | --epoch 50, --avg 22, --max-duration 500 |
+| fast beam search (set as default)  | 6.30    ß   | 5.89       | --epoch 50, --avg 22, --max-duration 500 |
+
+A pre-trained model and decoding logs can be found at .
```
Comment (pkufool) [inline egs/tedlium3/ASR/RESULTS.md:107]: Also here.

**Comment by desh2608** [comment]
> @desh2608 can you see whether there a big differences in ins/del ratios, and whether there are particular types of words that are most affected. Also try to compare your tensorflow plots for the 2 expts, so we can see if anything anomalous is happening.
> 
> BTW I think our modified RNN-T is equivalent to monotonic RNN-T: https://ieeexplore.ieee.org/document/9003822

@danpovey Here is the WER breakdown on the dev set using modified beam search:

| rnnt_type | Ins  | Del  | Sub  | WER  |
|-----------|------|------|------|------|
| regular   | 1.03 | 2.21 | 3.30 | 6.54 |
| modified  | 0.93 | 2.00 | 3.24 | 6.16 |

It seems there is improvement in all the error categories, so probably the model has just converged better? The tensorboard for both models are here: https://huggingface.co/desh2608/icefall-asr-tedlium3-zipformer/tensorboard

Yeah, it's the same as Monotonic RNNT. Niko Moritz (who is now at Meta) did a study on large scale training with them (https://ieeexplore-ieee-org.proxy1.library.jhu.edu/abstract/document/10022569) but it doesn't seem to help much. Perhaps it is only useful for small data to get better alignments at training time.

**Comment by desh2608** [comment]
@pkufool thanks for the pointer.

I had planned to add the missing links and fix the table, but I may have forgot. I will create a new PR with the fix it. For some reason, the decoding with beam search is stuck (although it ran file with the original model).

**Comment by pkufool** [comment]
> @pkufool thanks for the pointer.
> 
> I had planned to add the missing links and fix the table, but I may have forgot. I will create a new PR with the fix it. For some reason, the decoding with beam search is stuck (although it ran file with the original model).

OK，then you can delete beam search row, it is  super slow.

**Comment by danpovey** [comment]
> > @desh2608 can you see whether there a big differences in ins/del ratios, and whether there are particular types of words that are most affected. Also try to compare your tensorflow plots for the 2 expts, so we can see if anything anomalous is happening.
> > BTW I think our modified RNN-T is equivalent to monotonic RNN-T: https://ieeexplore.ieee.org/document/9003822
> 
> @danpovey Here is the WER breakdown on the dev set using modified beam search:
> 
> rnnt_type	Ins	Del	Sub	WER
> regular	1.03	2.21	3.30	6.54
> modified	0.93	2.00	3.24	6.16
> It seems there is improvement in all the error categories, so probably the model has just converged better? The tensorboard for both models are here: https://huggingface.co/desh2608/icefall-asr-tedlium3-zipformer/tensorboard
> 
> Yeah, it's the same as Monotonic RNNT. Niko Moritz (who is now at Meta) did a study on large scale training with them (https://ieeexplore-ieee-org.proxy1.library.jhu.edu/abstract/document/10022569) but it doesn't seem to help much. Perhaps it is only useful for small data to get better alignments at training time.

From your plots, it seems that the pruned loss is basically identical, but the simple loss is consistently better.
Yes, possibly it's about small data.  We'll have to see based on future experience; we should remember to try the option in future.

**Comment by pkufool** [comment]
@danpovey  I will try the `modified transducer` on 10000 hours Chinese dataset when there are free GPUs.

**Comment by danpovey** [comment]
> @danpovey I will try the `modified transducer` on 10000 hours Chinese dataset when there are free GPUs.

mm OK... I'd rather try on smaller amounts of data first, unless there is some indication that it might make more difference for larger data (which I doubt).

**Comment by pkufool** [comment]
@danpovey  We did not see such improvements (by using modified transduser) in librispeech dataset (reworked conformer model).  

>  I'd rather try on smaller amounts of data first, unless there is some indication that it might make more difference for larger data (which I doubt).

Then I will try S and M datasets first.
